File: pkg/datagen/datagen.go
// Package datagen procedurally generates DittoBench tool-calling datasets from
// templates using a seeded PRNG. The same seed yields a deterministic dataset;
// different seeds yield different prompts. This lets miners practice the
// run+score loop offline against fresh-looking cases.
//
// NOTE: the on-chain validator uses its own (held-out) generator. This is a
// practice generator only — do not overfit to it.
package datagen

import (
	"fmt"
	"math/rand"
	"time"

	"github.com/ditto-assistant/dittobench-starter-kit/pkg/protocol"
)

// category describes one generation template.
type category struct {
	name     string
	tools    []string // expected tool names (empty = abstention/no-tool)
	maxCalls int
	build    func(r *rand.Rand) (prompt string, behavior string)
}

// Word pools for randomized entities/phrasings.
var (
	topics    = []string{"quantum computing", "the French Revolution", "GraphQL", "sourdough starters", "Rust ownership", "the Voyager probes", "tariffs", "mitochondria", "the Bittensor subnet design", "Kubernetes operators"}
	people    = []string{"Ada", "Grace", "Linus", "Margaret", "Dennis", "Barbara", "Alan", "Katherine"}
	subjects  = []string{"my startup", "my dissertation", "the Q3 budget", "my marathon training", "the kitchen remodel", "my reading list", "the family reunion", "my side project"}
	sites     = []string{"https://example.com/post", "https://news.example.org/article", "https://blog.example.dev/entry", "https://docs.example.io/guide"}
	imageIdea = []string{"a neon city skyline at dusk", "a watercolor fox in a forest", "an isometric cozy bedroom", "a retro sci-fi book cover", "a minimalist mountain logo"}
	artifacts = []string{"README.md", "notes.txt", "plan.md", "todo.md", "config.yaml"}
	jobs      = []string{"refactor the auth module", "add unit tests for the parser", "migrate the config to TOML", "fix the flaky integration test", "scaffold a REST endpoint"}
	chitchat  = []string{"How are you doing today?", "Tell me a fun fact.", "What's a good name for a cat?", "Explain recursion simply.", "Give me a motivational quote."}
)

func pick(r *rand.Rand, pool []string) string { return pool[r.Intn(len(pool))] }

var categories = []category{
	{
		name: "memory_lookup", tools: []string{"search_memories"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("What did I say earlier about %s?", pick(r, subjects)),
				"Should call search_memories to recall prior user statements."
		},
	},
	{
		name: "memory_subject", tools: []string{"search_subjects"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Who is %s in my notes?", pick(r, people)),
				"Should call search_subjects to find the entity."
		},
	},
	{
		name: "web_search", tools: []string{"search_web"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("What's the latest news on %s?", pick(r, topics)),
				"Should call search_web for fresh information."
		},
	},
	{
		name: "link_read", tools: []string{"read_links"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Summarize this page for me: %s", pick(r, sites)),
				"Should call read_links on the provided URL."
		},
	},
	{
		name: "image_create", tools: []string{"create_image"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Make me an image of %s.", pick(r, imageIdea)),
				"Should call create_image with a prompt."
		},
	},
	{
		name: "artifacts_create", tools: []string{"artifacts"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Create a file called %s with my project outline.", pick(r, artifacts)),
				"Should call artifacts with action=create."
		},
	},
	{
		name: "agent_job", tools: []string{"execute_agent_job"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Kick off an agent to %s.", pick(r, jobs)),
				"Should call execute_agent_job with the task."
		},
	},
	{
		name: "settings_theme", tools: []string{"set_theme"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			mode := pick(r, []string{"dark", "light"})
			return fmt.Sprintf("Switch my app to %s mode.", mode),
				"Should call set_theme."
		},
	},
	{
		name: "settings_effort", tools: []string{"set_reasoning_effort"}, maxCalls: 2,
		build: func(r *rand.Rand) (string, string) {
			lvl := pick(r, []string{"high", "low", "medium"})
			return fmt.Sprintf("Set reasoning effort to %s please.", lvl),
				"Should call set_reasoning_effort."
		},
	},
	{
		name: "no_tool", tools: nil, maxCalls: 0,
		build: func(r *rand.Rand) (string, string) {
			return pick(r, chitchat),
				"Pure conversation; the harness should NOT call any tool."
		},
	},
	{
		name: "abstention", tools: nil, maxCalls: 0,
		build: func(r *rand.Rand) (string, string) {
			return fmt.Sprintf("Just chatting — no need to look anything up. By the way, %s",
				pick(r, chitchat)),
				"The harness should abstain from tool use."
		},
	},
}

// Generate builds a deterministic dataset of n cases for the given seed.
func Generate(seed int64, n int) protocol.Dataset {
	r := rand.New(rand.NewSource(seed))
	ds := protocol.Dataset{
		Seed:        seed,
		GeneratedAt: time.Now().UTC().Format(time.RFC3339),
	}
	for i := 0; i < n; i++ {
		cat := categories[r.Intn(len(categories))]
		prompt, behavior := cat.build(r)
		c := protocol.ToolCase{
			ID:               fmt.Sprintf("%s-%04d", cat.name, i),
			Category:         cat.name,
			Prompt:           prompt,
			MaxToolCalls:     cat.maxCalls,
			AllowExtraTools:  false,
			ExpectedBehavior: behavior,
		}
		for _, tn := range cat.tools {
			c.ExpectedTools = append(c.ExpectedTools, protocol.ToolSpec{Name: tn})
		}
		ds.ToolCases = append(ds.ToolCases, c)
	}
	return ds
}

File: src/datagen.rs
//! Procedurally generates small, fresh, randomized DittoBench datasets.
//!
//! Generation is deterministic per seed (a given seed always yields the same
//! dataset) but varies widely across seeds. The practice loop rotates the seed
//! on every run so no two evaluations are identical — the anti-overfit property
//! of the off-chain practice loop. Mirrors the Go validator
//! `internal/datagen`, extended with synthetic memory cases.

use chrono::Utc;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

use crate::protocol::{Dataset, MemoryCase, SeedMemory, ToolCase, ToolSpec};

/// One kind of tool-calling case and how to render it.
struct Category {
    name: &'static str,
    /// Expected tool name; empty means "no tool".
    tool: &'static str,
    templates: &'static [&'static str],
}

const SUBJECTS: &[&str] = &[
    "my dentist appointment",
    "the project deadline",
    "Sarah's birthday",
    "my car insurance",
    "the meeting notes",
    "my flight to Tokyo",
    "the grocery list",
    "my gym schedule",
    "the wifi password",
    "my passport number",
    "the book recommendation",
    "my doctor's name",
];
const TOPICS: &[&str] = &[
    "quantum computing",
    "the 2024 Olympics",
    "best espresso machines",
    "rust vs go",
    "climate policy",
    "the stock market today",
    "sourdough recipes",
    "electric vehicles",
    "the James Webb telescope",
];
const URLS: &[&str] = &[
    "https://example.com/article",
    "https://news.site/story",
    "https://blog.dev/post",
    "https://docs.io/guide",
    "https://github.com/org/repo",
];
const IMAGE_PROMPTS: &[&str] = &[
    "a sunset over mountains",
    "a robot drinking coffee",
    "a futuristic city skyline",
    "a watercolor fox",
    "an astronaut on a beach",
    "a neon cyberpunk street",
];
const ARTIFACT_KINDS: &[&str] = &[
    "a landing page",
    "a todo app",
    "a snake game",
    "a markdown resume",
    "a pomodoro timer",
    "a budget tracker",
];
const AGENT_TASKS: &[&str] = &[
    "scrape the latest headlines",
    "summarize this PDF",
    "refactor the auth module",
    "generate unit tests",
    "build a CSV report",
    "deploy the staging branch",
];
const THEMES: &[&str] = &["dark", "light", "system", "midnight", "solarized"];
const CHITCHAT: &[&str] = &[
    "hey, how's it going?",
    "thanks, that was helpful!",
    "tell me a joke",
    "what's your favorite color?",
    "good morning!",
    "you're awesome",
    "lol nice",
];
const ABSTENTIONS: &[&str] = &[
    "what's the meaning of life?",
    "should I quit my job?",
    "do you love me?",
    "what will the weather be like next year?",
    "who will win the next election?",
    "what am I thinking right now?",
];

/// Order is stable so seeding is reproducible.
const CATEGORIES: &[Category] = &[
    Category {
        name: "memory_lookup",
        tool: "search_memories",
        templates: &[
            "What did I say about %s?",
            "Remind me about %s.",
            "Do you remember %s?",
            "Look up %s from my memories.",
        ],
    },
    Category {
        name: "memory_subject",
        tool: "search_subjects",
        templates: &[
            "What subjects do I have notes on related to %s?",
            "Find the topic that covers %s.",
            "Which of my subjects mention %s?",
        ],
    },
    Category {
        name: "web_search",
        tool: "search_web",
        templates: &[
            "Search the web for %s.",
            "What's the latest on %s?",
            "Find recent news about %s.",
            "Google %s for me.",
        ],
    },
    Category {
        name: "link_read",
        tool: "read_links",
        templates: &[
            "Read %s and summarize it.",
            "What does this page say: %s",
            "Open %s and tell me the main points.",
        ],
    },
    Category {
        name: "image_create",
        tool: "create_image",
        templates: &[
            "Generate an image of %s.",
            "Create a picture of %s.",
            "Draw me %s.",
        ],
    },
    Category {
        name: "artifacts_create",
        tool: "artifacts",
        templates: &[
            "Build me %s.",
            "Make %s I can preview.",
            "Create %s as an interactive artifact.",
        ],
    },
    Category {
        name: "agent_job",
        tool: "execute_agent_job",
        templates: &[
            "Run a background job to %s.",
            "Kick off an agent to %s.",
            "Dispatch a task to %s.",
        ],
    },
    Category {
        name: "settings",
        tool: "set_theme",
        templates: &["Switch to %s mode.", "Set my theme to %s.", "Change the app theme to %s."],
    },
    Category {
        name: "no_tool",
        tool: "",
        templates: &["%s"],
    },
    Category {
        name: "abstention",
        tool: "",
        templates: &["%s"],
    },
];

fn pick<'a>(rng: &mut StdRng, pool: &'a [&'a str]) -> &'a str {
    pool[rng.gen_range(0..pool.len())]
}

fn filler_for(rng: &mut StdRng, category: &str) -> &'static str {
    match category {
        "memory_lookup" | "memory_subject" => pick(rng, SUBJECTS),
        "web_search" => pick(rng, TOPICS),
        "link_read" => pick(rng, URLS),
        "image_create" => pick(rng, IMAGE_PROMPTS),
        "artifacts_create" => pick(rng, ARTIFACT_KINDS),
        "agent_job" => pick(rng, AGENT_TASKS),
        "settings" => pick(rng, THEMES),
        "no_tool" => pick(rng, CHITCHAT),
        "abstention" => pick(rng, ABSTENTIONS),
        _ => pick(rng, TOPICS),
    }
}

/// Single `%s` substitution, matching the Go templates.
fn render(template: &str, filler: &str) -> String {
    template.replacen("%s", filler, 1)
}

// --- Memory case generation -------------------------------------------------

/// (topic, question template, answer, fact-prompt, fact-response).
struct MemoryFact {
    /// Short label for the fact (used in generated case ids / debugging).
    topic: &'static str,
    question: &'static str,
    /// The keyword/phrase the answer must surface (scored via substring match).
    answer: &'static str,
    prompt: &'static str,
    response: &'static str,
}

const MEMORY_FACTS: &[MemoryFact] = &[
    MemoryFact {
        topic: "my dog's name",
        question: "What's my dog's name?",
        answer: "Biscuit",
        prompt: "Just adopted a golden retriever puppy and named him Biscuit.",
        response: "Congrats on Biscuit! Golden retrievers are wonderful companions.",
    },
    MemoryFact {
        topic: "my favorite coffee order",
        question: "What's my usual coffee order?",
        answer: "oat milk flat white",
        prompt: "My go-to coffee is an oat milk flat white, double shot.",
        response: "Noted: oat milk flat white, double shot.",
    },
    MemoryFact {
        topic: "my anniversary",
        question: "When is my wedding anniversary?",
        answer: "June 14",
        prompt: "My wedding anniversary is June 14, don't let me forget it.",
        response: "I'll remember: your anniversary is June 14.",
    },
    MemoryFact {
        topic: "my car",
        question: "What car do I drive?",
        answer: "blue Subaru Outback",
        prompt: "I drive a blue Subaru Outback, 2021 model.",
        response: "Got it, a 2021 blue Subaru Outback.",
    },
    MemoryFact {
        topic: "my allergy",
        question: "What am I allergic to?",
        answer: "peanuts",
        prompt: "Reminder: I'm allergic to peanuts, it's serious.",
        response: "Understood, you have a serious peanut allergy.",
    },
    MemoryFact {
        topic: "my gym goal",
        question: "What's my current fitness goal?",
        answer: "run a half marathon",
        prompt: "My goal this year is to run a half marathon in under two hours.",
        response: "Great goal: a sub-two-hour half marathon.",
    },
    MemoryFact {
        topic: "my favorite author",
        question: "Who is my favorite author?",
        answer: "Ursula K. Le Guin",
        prompt: "My favorite author is Ursula K. Le Guin, hands down.",
        response: "Noted, Ursula K. Le Guin is your favorite author.",
    },
    MemoryFact {
        topic: "my work project",
        question: "What project am I working on at work?",
        answer: "billing service migration",
        prompt: "At work I'm leading the billing service migration to the new platform.",
        response: "Got it, you lead the billing service migration.",
    },
    MemoryFact {
        topic: "my hometown",
        question: "Where did I grow up?",
        answer: "Asheville",
        prompt: "I grew up in Asheville, North Carolina.",
        response: "Noted: you grew up in Asheville, NC.",
    },
    MemoryFact {
        topic: "my preferred editor",
        question: "Which code editor do I prefer?",
        answer: "Neovim",
        prompt: "I do all my coding in Neovim with a custom config.",
        response: "Understood, Neovim is your editor of choice.",
    },
];

/// Generates a deterministic-per-seed dataset of `n_tool` tool-calling cases
/// and `n_mem` memory cases. `n_tool` is clamped to [1, 200]; `n_mem` to
/// [0, 50].
pub fn generate(seed: i64, n_tool: usize, n_mem: usize) -> Dataset {
    let n_tool = n_tool.clamp(1, 200);
    let n_mem = n_mem.min(50);
    let mut rng = StdRng::seed_from_u64(seed as u64);

    let mut tool_cases = Vec::with_capacity(n_tool);
    for i in 0..n_tool {
        let cat = &CATEGORIES[rng.gen_range(0..CATEGORIES.len())];
        let template = cat.templates[rng.gen_range(0..cat.templates.len())];
        let filler = filler_for(&mut rng, cat.name);
        let prompt = render(template, filler);

        let mut tc = ToolCase {
            id: format!("{}-{}-{:04}", cat.name, seed, i),
            category: cat.name.to_string(),
            prompt,
            expected_tools: Vec::new(),
            max_tool_calls: 1,
            allow_extra_tools: false,
            expected_behavior: String::new(),
        };

        if !cat.tool.is_empty() {
            tc.expected_tools = vec![ToolSpec {
                name: cat.tool.to_string(),
                ..ToolSpec::default()
            }];
            tc.expected_behavior = format!("call {} exactly once", cat.tool);
        } else {
            tc.expected_tools = Vec::new();
            tc.max_tool_calls = 0;
            tc.expected_behavior = if cat.name == "abstention" {
                "answer or abstain without calling any tool".to_string()
            } else {
                "respond conversationally without calling any tool".to_string()
            };
        }
        tool_cases.push(tc);
    }

    let mut memory_cases = Vec::with_capacity(n_mem);
    for i in 0..n_mem {
        let fact = &MEMORY_FACTS[rng.gen_range(0..MEMORY_FACTS.len())];
        // 1-3 seed memories: the answer-bearing fact plus 0-2 distractors so
        // retrieval has to discriminate.
        let extra = rng.gen_range(0..3usize);
        let mut seeds = Vec::with_capacity(1 + extra);
        let days = rng.gen_range(1..30i64);
        seeds.push(SeedMemory {
            prompt: fact.prompt.to_string(),
            response: fact.response.to_string(),
            days_ago: days,
        });
        for _ in 0..extra {
            let d = &MEMORY_FACTS[rng.gen_range(0..MEMORY_FACTS.len())];
            seeds.push(SeedMemory {
                prompt: d.prompt.to_string(),
                response: d.response.to_string(),
                days_ago: rng.gen_range(1..60i64),
            });
        }
        let slug: String = fact
            .topic
            .chars()
            .map(|c| if c.is_ascii_alphanumeric() { c } else { '-' })
            .collect();
        memory_cases.push(MemoryCase {
            id: format!("mem-{}-{}-{:04}", slug, seed, i),
            question: fact.question.to_string(),
            expected_answer: fact.answer.to_string(),
            seed_memories: seeds,
        });
    }

    Dataset {
        seed,
        generated_at: Utc::now().to_rfc3339(),
        tool_cases,
        memory_cases,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn determinism_same_seed_same_dataset() {
        let a = generate(42, 30, 10);
        let b = generate(42, 30, 10);
        assert_eq!(a.tool_cases, b.tool_cases);
        assert_eq!(a.memory_cases, b.memory_cases);
    }

    #[test]
    fn variety_different_seeds_differ() {
        let a = generate(1, 30, 10);
        let b = generate(2, 30, 10);
        // Vanishingly unlikely to be identical across 30 randomized cases.
        assert_ne!(a.tool_cases, b.tool_cases);
    }

    #[test]
    fn counts_and_clamping() {
        let ds = generate(7, 0, 0);
        assert_eq!(ds.tool_cases.len(), 1, "n_tool clamps up to 1");
        assert_eq!(ds.memory_cases.len(), 0);

        let big = generate(7, 500, 100);
        assert_eq!(big.tool_cases.len(), 200, "n_tool clamps down to 200");
        assert_eq!(big.memory_cases.len(), 50, "n_mem clamps down to 50");
    }

    #[test]
    fn memory_cases_have_answer_bearing_seed() {
        let ds = generate(99, 5, 8);
        for mc in &ds.memory_cases {
            assert!(!mc.question.is_empty());
            assert!(!mc.expected_answer.is_empty());
            assert!(
                (1..=3).contains(&mc.seed_memories.len()),
                "memory case must have 1-3 seeds, got {}",
                mc.seed_memories.len()
            );
        }
    }

    #[test]
    fn tool_cases_cover_no_tool_and_tool_categories() {
        // A reasonably sized dataset should include both kinds across seeds.
        let ds = generate(123, 60, 0);
        let has_tool = ds.tool_cases.iter().any(|c| !c.expected_tools.is_empty());
        let has_no_tool = ds.tool_cases.iter().any(|c| c.expected_tools.is_empty());
        assert!(has_tool && has_no_tool, "dataset should mix tool/no-tool cases");
    }
}
